src/Microsoft.CodeAnalysis.BannedApiAnalyzers/Core/RestrictedInternalsVisibleToAnalyzer.cs

```diff
@@ -0,0 +1,237 @@
+                        foreach (TypedConstant namespaceConstant in namespaceConstants)
+                        {
+                            if (namespaceConstant.Kind == TypedConstantKind.Primitive &&
+                                namespaceConstant.Value is string namespaceName)
+                            {
+                                namespaceBuilder.Add(namespaceName);
+                            }
+                        }
+                    }
+
+                    if (namespaceBuilder.Count > 0)
+                    {
+                        builder.Add(referencedAssemblySymbol, namespaceBuilder.ToImmutable());
+                    }
+                }
+
+                return builder.ToImmutable();
+            }
+
+            void VerifySymbol(ISymbol symbol, SyntaxNode node, Action<Diagnostic> reportDiagnostic)
+            {
+                if (symbol != null &&
+                    IsBannedSymbol(symbol))
+                {
+                    var bannedSymbolDisplayString = symbol.ToDisplayString(SymbolDisplayFormats.QualifiedTypeAndNamespaceSymbolDisplayFormat);
+                    var assemblyName = symbol.ContainingAssembly.Name;
+                    var restrictedNamespaces = string.Join(", ", restrictedInternalsVisibleToMap[symbol.ContainingAssembly]);
+                    var diagnostic = node.CreateDiagnostic(Rule, bannedSymbolDisplayString, assemblyName, restrictedNamespaces);
+                    reportDiagnostic(diagnostic);
+                }
+            }
+
+            bool IsBannedSymbol(ISymbol symbol)
+            {
+                // Check if the symbol belongs to an assembly to which this compilation has restricted internals access
+                // and it is an internal symbol.
+                if (!restrictedInternalsVisibleToMap.TryGetValue(symbol.ContainingAssembly, out var allowedNamespaces) ||
+                    symbol.GetResultantVisibility() != SymbolVisibility.Internal)
+                {
+                    return false;
+                }
+
+                // Walk up containing namespace chain to explicitly look for an allowed namespace
+                // with restricted internals access.
+                var currentNamespace = symbol.ContainingNamespace;
+                while (currentNamespace != null && !currentNamespace.IsGlobalNamespace)
+                {
+                    // Check if we have already computed whether this namespace is banned or not.
+                    if (namespaceToIsBannedMap.TryGetValue(currentNamespace, out var isBanned))
+                    {
+                        return isBanned;
+                    }
+
+                    // Check if this namespace is explicitly marked as allowed through restricted IVT.
+                    if (allowedNamespaces.Contains(currentNamespace.ToDisplayString()))
+                    {
+                        var banned = namespaceToIsBannedMap.GetOrAdd(currentNamespace, false);
```

Comment:
Should we add all containing namespaces up to the current as not banned as well?

Reply:
Yes, that seems like a good optimization! I will add that logic.